auto_submit/lib/request_handling/authentication.dart

```diff
@@ -0,0 +1,177 @@
+
+  /// Authenticates the specified [request] and returns the associated
+  /// [AuthenticatedContext].
+  ///
+  /// See the class documentation on [AuthenticationProvider] for a discussion
+  /// of the different types of authentication that are accepted.
+  ///
+  /// This will throw an [Unauthenticated] exception if the request is
+  /// unauthenticated.
+  Future<AuthenticatedContext> authenticate(Request request) async {
+    final Map<String, String> reqHeader = request.headers;
+    final bool isCron = reqHeader['X-Appengine-Cron'] == 'true';
+    final String? idTokenFromHeader = reqHeader['X-Flutter-IdToken'];
+    final ClientContext clientContext = clientContextProvider();
+
+    if (isCron) {
+      // Authenticate cron requests
+      return AuthenticatedContext(clientContext: clientContext);
+    } else if (idTokenFromHeader != null) {
+      TokenInfo token;
+      try {
+        token = await tokenInfo(request);
+      } on Unauthenticated {
+        token = await tokenInfo(request, tokenType: 'access_token');
+      }
+      return authenticateToken(token, clientContext: clientContext);
+    }
+
+    throw const Unauthenticated('User is not signed in');
+  }
+
+  /// Gets oauth token information. This method requires the token to be stored in
+  /// X-Flutter-IdToken header.
+  Future<TokenInfo> tokenInfo(Request request, {String tokenType = 'id_token'}) async {
+    final String? idTokenFromHeader = request.headers['X-Flutter-IdToken'];
+    final http.Client client = httpClientProvider();
+    try {
+      final http.Response verifyTokenResponse = await client.get(Uri.https(
+        'oauth2.googleapis.com',
+        '/tokeninfo',
+        <String, String?>{
+          tokenType: idTokenFromHeader,
+        },
+      ));
+
+      if (verifyTokenResponse.statusCode != HttpStatus.ok) {
+        /// Google Auth API returns a message in the response body explaining why
+        /// the request failed. Such as "Invalid Token".
+        log.fine('Token verification failed: ${verifyTokenResponse.statusCode}; ${verifyTokenResponse.body}');
+        throw const Unauthenticated('Invalid ID token');
+      }
+
+      try {
+        return TokenInfo.fromJson(json.decode(verifyTokenResponse.body) as Map<String, dynamic>);
+      } on FormatException {
+        throw InternalServerError('Invalid JSON: "${verifyTokenResponse.body}"');
+      }
+    } finally {
+      client.close();
+    }
+  }
+
+  Future<AuthenticatedContext> authenticateToken(TokenInfo token, {required ClientContext clientContext}) async {
+    // Authenticate as a signed-in Google account via OAuth id token.
+    final String clientId = await config.getOauthClientId();
+    if (token.audience != clientId && !token.email!.endsWith('@google.com')) {
+      log.warning('Possible forged token: "${token.audience}" (expected "$clientId")');
+      throw const Unauthenticated('Invalid ID token');
+    }
+
+    // if (token.hostedDomain != 'google.com') {
```

Comment:
Can you rm the dead code from this file?

Reply:
I'm so sorry for this. I added too many files and codes which made me felt overwhelmed, thus forgot this.